### PR TITLE
[ggml] Optimize Q4_0 for arm backend

### PR DESCRIPTION
**Pull request by jaemini-shin**
## Commits to be reviewed in this PR

<details><summary>[ggml] Optimize Q4_0 for arm backend</summary><br />

- Add nntr_quantize_mat_q8_0_4x8_opt / nntr_quantize_row_q8_0_opt:
  NEON activation quantization kernels that use narrowing stores
  (vmovn + vst1q_s8) instead of per-lane stores; output is
  byte-identical to the reference, 31-36% (mat) / 26% (row) faster on
  a prime core. x86 and fallback alias to the reference implementation
- Rework the GEMM path of __ggml_q4_0_4x8_q8_0_GEMM (omp): switch
  online activation quantization to the _opt kernels and parallelize
  it over 8-group chunks when M is large enough to amortize waking the
  pool; fold the M % 4 != 0 case into the main GEMM branch so the
  leftover rows share the same tiled compute
- Make the GEMM column chunk adaptive (16..64, targeting ~64 tasks per
  thread; row chunk stays 16) for balanced work stealing across
  asymmetric cores: measured +11% at M=256 and +4% at M=1024 end to
  end
- Add unittest_nntrainer_ggml_arm (aarch64 only): byte-identical
  checks for the _opt kernels including d == 0 and rounding-boundary
  edge cases, tiling-invariance tests for the GEMM/GEMV kernels, and
  cold-weight benchmarks that evict the packed RHS with dc civac and
  wake the thread pool right before every timed call, reporting
  min/median/mean/max/stdev with chronological samples

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>
Signed-off-by: Jaemin Shin <jaemin980311@gmail.com>

</details>

### Summary

This PR optimizes the Q4_0 GEMM for arm devices(neon).

### Performance on Galaxy S26+

* Before this PR
GEMM kernel performance: 20508.9 ± 328.147 us
qwen3-0.6b 1024 input prefill: 434.082 TPS

* From this PR
GEMM kernel performance: 18452.3 ± 212.768 us
qwen3-0.6b 1024 input prefill: 460.846 TPS

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>
Signed-off-by: Jaemin Shin <jaemin980311@gmail.com>

### CODE REVIEW

File: nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl.h [hunk 1]
```diff
@@ -64,6 +64,9 @@ void nntr_gemv_q4_K_8x8_q8_K(int n, float *__restrict s, size_t bs,
 void nntr_quantize_mat_q8_0_4x8(const float *__restrict x, void *__restrict vy,
                                 int64_t k);
 
+void nntr_quantize_mat_q8_0_4x8_opt(const float *__restrict x,
+                                    void *__restrict vy, int64_t k);
```
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl.h]: This optimized function is about 10% faster than the reference code.

File: nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl.h [hunk 2]
```diff
@@ -98,6 +101,9 @@ size_t nntr_quantize_q8_0(const float *__restrict src, void *__restrict dst,
 void nntr_quantize_row_q8_0(const float *__restrict x, void *__restrict y,
                             int64_t k);
 
+void nntr_quantize_row_q8_0_opt(const float *__restrict x, void *__restrict y,
+                                int64_t k);
```
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl.h]: This optimized function is about 67% faster than the reference code.

File: nntrainer/tensor/cpu_backend/ggml_interface/ggml_interface_omp.cpp [hunk 3]
```diff
@@ -53,7 +53,7 @@ void __ggml_q4_0_4x8_q8_0_GEMM(const unsigned int M, const unsigned int N,
                               (void *)((char *)B + M_step_start * B_step),
                               QA.data(), M, M_step_end - M_step_start);
     });
-  } else if (M % 4 != 0) {
+  } else { // GEMM
```
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/ggml_interface/ggml_interface_omp.cpp:56]: ### Before this PR
GEMV, GEMM (m%4 == 0), GEMM

### From this PR
GEMV, GEMM

**Comment by Jungwon-Lee** [comment]
@jaemini-shin It looks great. It makes ggml quantization faster.
Could this replace `nntr_quantize_mat_q8_0_4x8 `directly, without introducing `nntr_quantize_mat_q8_0_4x8_opt`?

**Comment by jaemini-shin** [comment]
> @jaemini-shin It looks great. It makes ggml quantization faster. Could this replace `nntr_quantize_mat_q8_0_4x8 `directly, without introducing `nntr_quantize_mat_q8_0_4x8_opt`?

@Jungwon-Lee I'm working on refactoring ggml files, let me replace it in another PR!

File: nntrainer/tensor/cpu_backend/ggml_interface/ggml_interface_omp.cpp [hunk 4]
```diff
@@ -63,26 +63,44 @@ void __ggml_q4_0_4x8_q8_0_GEMM(const unsigned int M, const unsigned int N,
...
 
     // Compute 4-divisible-M row portion with multithreaded GEMM
+    // Row chunk is fixed at 16 (L1-friendly for K=1k-8k range).
+    // Column chunk scales to maintain ~64 tasks per thread for balanced
+    // work-stealing across asymmetric cores.
     unsigned int row_chunk_size = 16;
     size_t row_loop = (M4 * 4 + row_chunk_size - 1) / row_chunk_size;
+    size_t col_chunk_size = (unsigned int)std::clamp<size_t>(
+      (size_t)N * row_loop / ((size_t)64 * tm.getComputeThreadCount()) / 4 * 4,
+      16, 64);
```
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/ggml_interface/ggml_interface_omp.cpp:101]: Now, column chunk is scaled by thread number and N.

**Review by Jungwon-Lee — APPROVED** [review]
LGTM

**Comment by jaemini-shin** [comment]
@Jungwon-Lee I found additional points that could be optimized more. I'll include every change in this PR!

File: nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl_neon.cpp
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl_neon.cpp:1]: This optimized function is about 60% faster than the reference code.

File: nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl_quant.cpp
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl_quant.cpp:1]: This optimized function is about 40% faster than the reference code.

File: nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl_utils.h [hunk 5]
```diff
@@ -36,6 +36,22 @@
 #include <immintrin.h>
 #endif // !defined(__aarch64__)
 
+// Inline fp32 -> fp16 conversion for the hot quantization loops.
+// TODO: fold this into nntr_fp32_to_fp16 (make it inline) and drop this helper.
+#if defined(__ARM_NEON) && defined(ENABLE_FP16) &&                             \
+  !(defined(__CUDACC__) && __CUDACC_VER_MAJOR__ <= 11) && !defined(__MUSACC__)
+static inline nntr_fp16_t nntr_fp32_to_fp16_inline(float f) {
```
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/ggml_interface/nntr_ggml_impl/nntr_ggml_impl_utils.h:43]: This temporary helper is going to replace `nntr_fp32_to_fp16`.

**Review by lhs8928 — APPROVED** [review]
LGTM

**Review by jayden0701 — APPROVED** [review]
Great! LGTM!

**Comment by jijoongmoon** [comment]
I think we can rename the *omp* file since we are  not going to use the openmp anymore.

**Comment by jaemini-shin** [comment]
> I think we can rename the _omp_ file since we are not going to use the openmp anymore.

@jijoongmoon I'm already working on refactoring ggml directory, you can find it soon!

**Review by jijoongmoon — APPROVED** [review]
LGTM
